README.md
[![CS & Lint](https://github.com/Automattic/fb-instant-articles/actions/workflows/cs-lint.yml/badge.svg)](https://github.com/Automattic/fb-instant-articles/actions/workflows/cs-lint.yml) [![Integration Tests](https://github.com/Automattic/fb-instant-articles/actions/workflows/tests.yml/badge.svg)](https://github.com/Automattic/fb-instant-articles/actions/workflows/tests.yml) [![Stale monitor](https://github.com/Automattic/fb-instant-articles/actions/workflows/stale.yml/badge.svg)](https://github.com/Automattic/fb-instant-articles/actions/workflows/stale.yml)
# Instant Articles for WP

Enable [Instant Articles for Facebook](https://developers.facebook.com/docs/instant-articles) on your WordPress site.

## Maintenance 🚩
Support for this plugin is provided for PHP 8 migrations, but both Automattic/WPVIP and Facebook/Meta have stopped all active development. At this point in the product lifecycle, publishers should be advised to stop using this plugin, and instead, use direct site traffic for Facebook content.
Meta have announced that Instant Articles will have all support dropped at April 2023.

## Description

This plugin adds support for Instant Articles for Facebook, which is a new way for publishers to distribute fast, interactive stories on Facebook. Instant Articles are preloaded in the Facebook mobile app so they load instantly.

With the plugin active, a new menu will be available for you to connect to your Facebook Page and start publishing your Instant Articles. You'll also see the status of each Instant Articles submission on the edit page of your posts.

A best effort is made to generate valid Instant Article markup from your posts' content/metadata and publish it to Facebook. The plugin knows how to transform your posts' markup from a set of rules which forms a mapping between elements in you *source markup* and the valid *Instant Article components*. We refer to this “glue” between the two as the ***Transformer Rules***.

Built-in to the plugin are many [pre-defined transformer rules](https://github.com/Automattic/facebook-instant-articles-wp/blob/master/rules-configuration.json) which aims to cover standard WordPress installations. If your WordPress content contains elements which are not covered by the built-in ruleset, you can define your own additional rules to extend or override the defaults in the Settings of this plugin, under: **Plugin Configuration** > **Publishing Settings** > **Custom transformer rules**.

## Access to Instant Articles

The current criteria for access to Instant Articles are:

- your Facebook Page must have an established presence of at least 90 days
- your content adheres to the [Instant Article Policies](https://developers.facebook.com/docs/instant-articles/policy/)

Before your Instant Articles can be published on Facebook, your feed must undergo an initial review and approval. Facebook requires a minimum number of 10 articles in your feed before being eligible for review. The review process checks that your draft Instant Articles are properly formatted, have content consistency with their mobile web counterparts, and adhere to the [community standards](https://www.facebook.com/communitystandards/) and [content policies](https://www.facebook.com/help/publisher/1348682518563619).

It's important to note that if you use meta fields to add extra text, images or videos to your Posts, Facebook will expect you to add these to your Instant Articles output too. This plugin includes hooks to help you do that.

[See Facebook's documentation for full details of the submission process.](https://developers.facebook.com/docs/instant-articles)

Once your feed has been approved, new posts will automatically be taken live on Instant Articles, and existing posts will be taken live once you update them. In order to stay in the Instant Articles program, you must remain active by creating new content and maintaining a minimal readership.

## Installation

1. Run `composer install` on the root of the plugin folder. Make sure you have [Composer](https://github.com/composer/composer) installed.
2. Upload the folder to the `/wp-content/plugins/` directory
3. Activate the plugin through the 'Plugins' menu in WordPress

## Frequently Asked Questions

**Why is there content from my post missing in the generated Instant Article?**

More likely than not, this is because there is markup in the body of your post that is not mapped to a recognized Instant Article component. On the “Edit Post” screen for your post, look for additional information about the *transformed* output shown within the **Facebook Instant Articles** module located at the bottom of the screen.

**Why doesn't my post appear in the list of Instant Articles in the publisher tools?**
Your posts are imported to your library when they are shared on Facebook for the first time.

Alternatively, you can trigger a manual scrape by pasting your URL on our [Share Debugger](http://developers.facebook.com/tools/debug)

Only Instant Articles with URLs in [domains you have claimed](https://developers.facebook.com/docs/instant-articles/guides/publishertools#connect) will show up in the Publishing Tools section.
**In the Instant Articles module for my post, what does the “This post was transformed into an Instant Article with some warnings” message mean?**

When transforming your post into an Instant Article, this plugin will show warnings when it encounters content which might not be valid when published to Facebook. When you see this message, it is recommended to resolve each warning individually.

**What does the “No rules defined for ____ in the context of  ____” warning mean?**

This plugin transforms your post into an Instant Article by matching markup in your content to one of the [components available](https://github.com/facebook/facebook-instant-articles-sdk-php/blob/master/docs/QuickStart.md#transformer-classes) in Instant Articles markup. Although the plugin contains many [built-in rules](https://github.com/Automattic/facebook-instant-articles-wp/blob/master/rules-configuration.json) to handle common cases, there may be markup in your post which is not recognized by these existing rules. In this case, you may be required to define some of your own rules. See below for more details about where and how.

**How do I define my own transformer rules so that content from my site is rendered appropriately in an Instant Article?**

Your custom rules can be defined in the Settings of this plugin, under: **Plugin Configuration** > **Publishing Settings** > **Custom transformer rules**. More detailed instructions about all the options available is documented in the [Custom Transformer Rules](https://github.com/facebook/facebook-instant-articles-sdk-php/blob/master/docs/QuickStart.md#custom-transformer-rules) section of the Facebook Instant Articles SDK.

**I know of a custom transformer rule which is pretty common in the community. How can it be included by default in the plugin?**

You can propose popular transformer rules to be included in the plugin by [suggesting it on GitHub](https://github.com/Automattic/facebook-instant-articles-wp/issues/new).

**How do I post articles to Instant Articles after plugin is installed?**

In order to import your posts to your Instant Articles library on Facebook you need to [Connect Your Site](https://developers.facebook.com/docs/instant-articles/guides/publishertools#connect) first. Then either share your posts on your page or use the [Sharing Debugger](https://developers.intern.facebook.com/tools/debug/sharing/) to scrape them. After you have 10 articles imported, you will be able to submit them for review.

**How do I change the feed slug/URL if I'm using the RSS integration?**

To change the feed slug, set the constant INSTANT_ARTICLES_SLUG to whatever you like. If you do, remember to flush the rewrite rules afterwards.
By default it is set to `instant-articles` which usually will give you a feed URL set to `/feed/instant-articles`

**How do I flush the rewrite rules after changing the feed slug?**

Usually simply visiting the permalinks settings page in the WordPress dashboard will do the trick (/wp-admin/options-permalink.php)

## Changelog

Please visit the [changelog](https://github.com/Automattic/fb-instant-articles/blob/develop/CHANGELOG.md).


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [5.0.2] - 2022-10-29

- Fix: Make `Instant_Articles_Post::$instant_article` property public.

## [5.0.1] - 2022-10-18

- Refresh dependency that handles fixing of transformer encodings.

## [5.0.0] - 2022-10-18

**Minimum PHP requirement is increased from PHP 5.4 to PHP 7.1.**
**Minimum WordPress requirement is increased from WP 4.3.0 to WP 4.7.0.**

- [#683](https://github.com/Automattic/fb-instant-articles/pull/683) Fix wizard option dynamic properties for PHP 8.2.
- [#831](https://github.com/Automattic/fb-instant-articles/pull/831) Fix the way filters are applied in get_cover_media().
- [#1025](https://github.com/Automattic/fb-instant-articles/pull/1025) Support for installing as Composer Package.
- [#1058](https://github.com/Automattic/fb-instant-articles/pull/1058) Allow post content caching to be disabled.
- [#1073](https://github.com/Automattic/fb-instant-articles/pull/1073) Update release.sh.
- [#1076](https://github.com/Automattic/fb-instant-articles/pull/1076) Reduce plugin's cache burden for remote requests.
- [#1082](https://github.com/Automattic/fb-instant-articles/pull/1082) Fix: pass the missing $post param to the get_the_excerpt filter.
- [#1092](https://github.com/Automattic/fb-instant-articles/pull/1092) Add support for PHP 8.
- [#1095](https://github.com/Automattic/fb-instant-articles/pull/1095) Add GitHub workflow to mark issues and PR as stale.
- [#1097](https://github.com/Automattic/fb-instant-articles/pull/1097) Actions: Remove separate PHP 8.2 composer install.
- [#1098](https://github.com/Automattic/fb-instant-articles/pull/1098) Improve tests.
- [#1099](https://github.com/Automattic/fb-instant-articles/pull/1099) Basic code maintenance improvements.

## [4.2.1] - 2020-06-10
- [#1049](https://github.com/automattic/fb-instant-articles/pull/1049) Add support for new guttenberg layout elements (@diegoquinteiro)
- [#1012](https://github.com/automattic/fb-instant-articles/pull/1012) update composer modules (@paulschreiber)
- [#1001](https://github.com/automattic/fb-instant-articles/pull/1001) Load functions which rely on `plugins_loaded` action hook in VIP Go environment (@paulschreiber)
- [#990](https://github.com/automattic/fb-instant-articles/pull/990) avoid undefined index error in $display_warning_column (@paulschreiber)
- [#1002](https://github.com/automattic/fb-instant-articles/pull/1002) add Comscore plugin (@paulschreiber)
- [#1007](https://github.com/automattic/fb-instant-articles/pull/1007) Update Facebook SDK versions (@paulschreiber)

## [4.2.0] - 2018-11-29
- [#997](https://github.com/automattic/fb-instant-articles/pull/997) Removed fields from deprecated feature to enable comments/likes/share from media content (@everton-rosario)

## [4.1.1] - 2018-07-17
- [#962](https://github.com/automattic/fb-instant-articles/pull/962) Updating description for the plugin on website (@Blakomen)
- [#1](https://github.com/automattic/fb-instant-articles/pull/1) Merging into local (@diegoquinteiro, @everton-rosario, @algmelo)
- [#957](https://github.com/automattic/fb-instant-articles/pull/957) Delete current transient with autoload=yes. (@everton-rosario, @algmelo)

## [4.1.0] - 2018-07-02
- [#937](https://github.com/automattic/fb-instant-articles/pull/937) Add token-less re-scrape with request signature (@diegoquinteiro)
- [#952](https://github.com/automattic/fb-instant-articles/pull/952) Updating description for the plugin (@Blakomen)
- [#910](https://github.com/automattic/fb-instant-articles/pull/910) Add PassThroughRule for Grammarly (@robbiet480)
- [#941](https://github.com/automattic/fb-instant-articles/pull/941) Add support for GTM4WP (@robbiet480)
- [#898](https://github.com/automattic/fb-instant-articles/pull/898) Added tranformation warning indicator column (@Kluny)
- [#904](https://github.com/automattic/fb-instant-articles/pull/904) Enables setup for updating/invalidating articles cache from Facebook scraper (@everton-rosario, @diegoquinteiro)
- [#896](https://github.com/automattic/fb-instant-articles/pull/896) feat(playbuzz): support for the next major playbuzz plugin version (@playbuzz)
- [#740](https://github.com/automattic/fb-instant-articles/pull/740) (wizard) Update wording about AMP markup gen (@Djiit)
- [#870](https://github.com/automattic/fb-instant-articles/pull/870) [bugfix] calling get_article_style to apply the style filter (@vkama)

## [4.0.6] - 2017-12-04
- [#814](https://github.com/automattic/fb-instant-articles/pull/814) Add default rule for Twitter blockquote (@pestevez)
- [#806](https://github.com/automattic/fb-instant-articles/pull/806) Clarify requirement of site connection in README (@pestevez)
- [#798](https://github.com/automattic/fb-instant-articles/pull/798) remove article rescrape code (@timjacobi)
- [#777](https://github.com/automattic/fb-instant-articles/pull/777) update instructions (@timjacobi)
- [#778](https://github.com/automattic/fb-instant-articles/pull/778) Add label definitions (@timjacobi)
- [#792](https://github.com/automattic/fb-instant-articles/pull/792) Change meta box and readme so user doesn't think IAs are actually submitted to Facebook (@timjacobi)
- [#797](https://github.com/automattic/fb-instant-articles/pull/797) Display error message if meta box can't be loaded (@timjacobi)
- [#775](https://github.com/automattic/fb-instant-articles/pull/775) Adds caching to instant_articles_embed_oembed_html (@emrikol)

## [4.0.5] - 2017-08-24
- [#750](https://github.com/automattic/fb-instant-articles/pull/750) Fix query limit and escaping on AMP generation (@diegoquinteiro)

## [4.0.4] - 2017-07-27
- [#738](https://github.com/automattic/fb-instant-articles/pull/738) Add changelog to script (@diegoquinteiro)

instant-articles.php
<?php
/**
 * Plugin Name: Instant Articles for WP
 * Description: Add support for Instant Articles for Facebook to your WordPress site.
 * Author: Automattic, Dekode
 * Author URI: https://vip.wordpress.com/plugins/instant-articles/
 * Version: 0.1
 * Text Domain: instant-articles
 * License: GPLv2
 * License URI: http://www.gnu.org/licenses/gpl-2.0.html
 */

defined( 'ABSPATH' ) || die('Shame on you');


// Let users define their own feed slug
if ( ! defined( 'INSTANT_ARTICLES_SLUG' ) ) {
	define( 'INSTANT_ARTICLES_SLUG', 'instant-articles' );
}

require_once( dirname( __FILE__ ) . '/dom-transform-filters/class-instant-articles-dom-transform-filter.php' );
require_once( dirname( __FILE__ ) . '/class-instant-articles-dom-transform-filter-runner.php' );
require_once( dirname( __FILE__ ) . '/class-instant-articles-post.php' );

/**
 * Plugin activation hook to add our rewrite rules
 *
 * @since 0.1
 */
function instant_articles_activate(){
	instant_articles_init();
	flush_rewrite_rules();
}
register_activation_hook( __FILE__, 'instant_articles_activate' );

/**
 * Plugin activation hook to remove our rewrite rules
 *
 * @since 0.1
 */
function instant_articles_deactivate(){
	flush_rewrite_rules();
}
register_deactivation_hook( __FILE__, 'instant_articles_deactivate' );

/**
 * Load plugin textdomain
 *
 * @since 0.1
 */
function instant_articles_load_textdomain() {
	load_plugin_textdomain( 'instant-articles', false, plugin_dir_path( __FILE__ ) . '/languages' );
}
add_action( 'plugins_loaded', 'instant_articles_load_textdomain' );

/**
 * Register our special feed
 *
 * @since 0.1
 */
function instant_articles_init() {
	add_feed( INSTANT_ARTICLES_SLUG, 'instant_articles_feed' );
}
add_action( 'init', 'instant_articles_init' );

/**
 * Feed display callback
 *
 * @since 0.1
 */
function instant_articles_feed() {

	// Load compat layers
	include( dirname( __FILE__ ) . '/compat.php' );

	// Load shortcode handlers
	include( dirname( __FILE__ ) . '/shortcodes.php' );

	// Load embedded content handlers
	include( dirname( __FILE__ ) . '/embeds.php' );

	// Load the feed template
	include( dirname( __FILE__ ) . '/feed-template.php' );

}

/**
 * Register included DOM transformation filters
 *
 * @since 0.1
 */
function instant_articles_register_transformation_filters() {

	include( dirname( __FILE__ ) . '/dom-transform-filters/class-instant-articles-dom-transform-filter-video.php' );
	Instant_Articles_DOM_Transform_Filter_Runner::register( 'Instant_Articles_DOM_Transform_Filter_Video' );

	include( dirname( __FILE__ ) . '/dom-transform-filters/class-instant-articles-dom-transform-filter-image.php' );
	Instant_Articles_DOM_Transform_Filter_Runner::register( 'Instant_Articles_DOM_Transform_Filter_Image' );

	include( dirname( __FILE__ ) . '/dom-transform-filters/class-instant-articles-dom-transform-filter-blockquote.php' );
	Instant_Articles_DOM_Transform_Filter_Runner::register( 'Instant_Articles_DOM_Transform_Filter_Blockquote' );

	include( dirname( __FILE__ ) . '/dom-transform-filters/class-instant-articles-dom-transform-filter-unordered-list.php' );
	Instant_Articles_DOM_Transform_Filter_Runner::register( 'Instant_Articles_DOM_Transform_Filter_Unordered_List' );

	include( dirname( __FILE__ ) . '/dom-transform-filters/class-instant-articles-dom-transform-filter-ordered-list.php' );
	Instant_Articles_DOM_Transform_Filter_Runner::register( 'Instant_Articles_DOM_Transform_Filter_Ordered_List' );

	include( dirname( __FILE__ ) . '/dom-transform-filters/class-instant-articles-dom-transform-filter-table.php' );
	Instant_Articles_DOM_Transform_Filter_Runner::register( 'Instant_Articles_DOM_Transform_Filter_Table' );

	include( dirname( __FILE__ ) . '/dom-transform-filters/class-instant-articles-dom-transform-filter-address.php' );
	Instant_Articles_DOM_Transform_Filter_Runner::register( 'Instant_Articles_DOM_Transform_Filter_Address' );

	//Instant articles only support h1 and h2. Convert h3-h6 to h2.
	include( dirname( __FILE__ ) . '/dom-transform-filters/class-instant-articles-dom-transform-filter-heading.php' );
	Instant_Articles_DOM_Transform_Filter_Runner::register( 'Instant_Articles_DOM_Transform_Filter_Heading' );

	// Remove empty elements
	include( dirname( __FILE__ ) . '/dom-transform-filters/class-instant-articles-dom-transform-filter-emptyelements.php' );
	Instant_Articles_DOM_Transform_Filter_Runner::register( 'Instant_Articles_DOM_Transform_Filter_Emptyelements', 90 );
}
add_action( 'instant_articles_register_dom_transformation_filters', 'instant_articles_register_transformation_filters' );


/**
 * Modify the main query for our feed.
 *
 * We want the posts in the modified order, to provide Facebook with content updates even for older posts.
 * Facebook will only import 100 posts at the time.
 * Facebook will only update posts modified within the last 24 hours
 *
 * @param WP_Query  $query  The WP_Query object. Passed by reference.
 */
function instant_articles_query( $query ) {

	if ( $query->is_main_query() && $query->is_feed( INSTANT_ARTICLES_SLUG ) ) {

		$query->set( 'orderby', 'modified' );
		$query->set( 'posts_per_page', 100 );
		$query->set( 'posts_per_rss', 100 );

		/**
		 * If the constant INSTANT_ARTICLES_LIMIT_POSTS is set to true, we will limit the feed
		 * to only include posts which are modified within the last 24 hours.
		 * Facebook will initially need 100 posts to pass the review, but will only update
		 * already imported articles if they are modified within the last 24 hours.
		 */
		if ( defined( 'INSTANT_ARTICLES_LIMIT_POSTS' ) && INSTANT_ARTICLES_LIMIT_POSTS ) {
			$query->set( 'date_query', array(
				array(
					'column' => 'post_modified',
					'after'  => '1 day ago',
				),
			) );
		}
	}

}
add_action( 'pre_get_posts', 'instant_articles_query', 10, 1 );


/**
 * Filter the SQL query to not include posts with empty content -- FB will complain
 *
 * @since 0.1
 * @param string   $where  The original where part of the SQL statement
 * @param WP_Query $query  The WP_Query instance
 * @return string  The modified where part of the SQL statement
 */
function instant_articles_query_where( $where, $query ) {

	// Don’t modify the SQL query with a potentially expensive WHERE clause if we’re OK with fewer posts than 100 and are OK with filtering in the loop
	if ( defined( 'INSTANT_ARTICLES_LIMIT_POSTS' ) && INSTANT_ARTICLES_LIMIT_POSTS ) {
		return $where;
	}

	if ( $query->is_main_query() && $query->is_feed( INSTANT_ARTICLES_SLUG ) ) {
		global $wpdb;
		$where .= " AND {$wpdb->posts}.post_content NOT LIKE ''";
	}
	return $where;

}
add_filter( 'posts_where' , 'instant_articles_query_where', 10, 2 );


